core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -273,20 +273,23 @@ class Partition(val topic: String,
   }
 
   /**
-   * Check and maybe expand the ISR of the partition.
+   * Check and maybe expand the ISR of the partition. A replica is in ISR of the partition if and only if
+   * replica's LEO >= HW and replica's lag <= replicaLagTimeMaxMs.
    *
    * This function can be triggered when a replica's LEO has incremented
    */
-  def maybeExpandIsr(replicaId: Int) {
+  def maybeExpandIsr(replicaId: Int, logReadResult: LogReadResult) {
     val leaderHWIncremented = inWriteLock(leaderIsrUpdateLock) {
       // check if this replica needs to be added to the ISR
       leaderReplicaIfLocal() match {
         case Some(leaderReplica) =>
           val replica = getReplica(replicaId).get
-          val leaderHW = leaderReplica.highWatermark
           if(!inSyncReplicas.contains(replica) &&
              assignedReplicas.map(_.brokerId).contains(replicaId) &&
-                  replica.logEndOffset.offsetDiff(leaderHW) >= 0) {
+             // This approximates the requirement logReadResult.fetchTimeMs - replica.lastCaughtUpTimeMs <= replicaManager.config.replicaLagTimeMaxMs.
+             // We don't directly specify the above requirement in order to make maybeExpandIsr() consistent with ReplicaFetcherThread.shouldFollowerThrottle()
+             // A offset replica whose lag > ReplicaFetcherThread may still exceed hw because maybeShrinkIsr() is called periodically
+             logReadResult.info.fetchOffsetMetadata.messageOffset >= logReadResult.hw) {
```

Comment:
I think we should use leaderReplica.highWatermark instead of logReadResult.hw since the former is more precise. Actually, could we just keep the original code?